File: Python/Q7_Associacao.py
class Escola:
    def __init__(self,nome):
        self.nome = nome
        self.professores = []
    def __str__(self):
        return f"{self.nome}{self.professores}"
class Professor:
    def __init__(self,nome,*escolas):
        self.nome = nome
        self.escolas = escolas
        for i in range(len(escolas)):
            escolas[i].professores.append(self)
    
    def __str__(self) -> str:
        return f"{self.nome}({self.escolas})"

escola = Escola("Ideal")
escola2 = Escola("Iouta")
professor = Professor("Zé",escola,escola2)
professor2 = Professor("João",escola,escola2)
print(escola2.professores)

File: go/Q7_Associacao/Q7_Associacao.go
package main

import (
	"fmt"
)

type Professor struct {
	nome      string
	escolas   []*Escola 
}

type Escola struct {
	nome      string
	professores []*Professor 
}

func (e *Escola) AdicionarProfessor(p *Professor) {
	e.professores = append(e.professores, p)
	p.escolas = append(p.escolas, e)
}

func (e Escola) Info() string {
	return fmt.Sprintf("Escola: %s, Professores: %d", e.nome, len(e.professores))
}

func (p Professor) Info() string {
	return fmt.Sprintf("Professor: %s, Escolas: %d", p.nome, len(p.escolas))
}

func main() {
	prof1 := &Professor{nome: "João"}
	prof2 := &Professor{nome: "Maria"}

	escola1 := Escola{nome: "Escola A"}
	escola2 := Escola{nome: "Escola B"}

	escola1.AdicionarProfessor(prof1)
	escola1.AdicionarProfessor(prof2)
	escola2.AdicionarProfessor(prof1)

	fmt.Println(escola1.Info())
	fmt.Println(escola2.Info())

	fmt.Println(prof1.Info())
	fmt.Println(prof2.Info())
}
